File: src/main/java/com/leet/easy/SquaresOfASortedArray.java
package com.leet.easy;

import java.util.Arrays;

public class SquaresOfASortedArray {
  public static int[] sortedSquares(int[] nums) {
    int[] squares = new int[nums.length];
    int index = nums.length - 1;

    int low = 0, high = nums.length - 1;

    while (low <= high) {
      int lowItemSquare = nums[low] * nums[low];
      int highItemSquare = nums[high] * nums[high];
      if (lowItemSquare > highItemSquare) {
        squares[index--] = lowItemSquare;
        low++;
      } else {
        squares[index--] = highItemSquare;
        high--;
      }
    }
    return squares;
  }

  public static void main(String[] args) {
    int[] nums = new int[] { -4, -1, 0, 3, 10 };
    System.out.println(Arrays.toString(sortedSquares(nums)));
  }
}

File: src/main/java/com/leet/easy/SquaresOfASortedArray.rs
struct Solution {}

impl Solution {
  pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
    let mut low = 0;
    let mut high = nums.len() - 1;
    let mut ans = vec![0; nums.len()];
    let mut pos = nums.len() - 1;

    while low < high {
      if nums[low].abs() > nums[high].abs() {
        ans[pos] = nums[low] * nums[low];
        low += 1;
      } else {
        ans[pos] = nums[high] * nums[high];
        high -= 1;
      }
      pos -= 1;
    }
    ans[pos] = nums[low] * nums[low];
    ans
  }
}

fn main() {
  let ans = Solution::sorted_squares(vec![-7, -3, 2, 3, 11]);
  println!("{:?}", ans);
}
